azure-iot-device/tests/common/test_async_adapter.py

```diff
@@ -65,50 +65,91 @@ def some_function():
-        callback = async_adapter.AwaitableCallback(mock_function)
+    async def test_calling_object_completes_future(self):
+        callback = async_adapter.AwaitableCallback()
         assert not callback.future.done()
         callback()
         await asyncio.sleep(0.1)  # wait to give time to complete the callback
         assert callback.future.done()
+        assert not callback.future.exception()
+        await callback.completion()
 
-    @pytest.mark.it("Can be called using positional arguments")
-    async def test_can_be_called_using_positional_args(self, mocker, mock_function):
-        callback = async_adapter.AwaitableCallback(mock_function)
-        result = callback(1, 2, 3)
-        assert mock_function.call_count == 1
-        assert mock_function.call_args == mocker.call(1, 2, 3)
-        assert result == mock_function.return_value
+    @pytest.mark.it(
+        "Completes the instance Future when a call is invoked on the instance (with return_arg_name)"
+    )
+    async def test_calling_object_completes_future_with_return_arg_name(
+        self, fake_return_arg_name, fake_return_arg_value
+    ):
+        callback = async_adapter.AwaitableCallback(return_arg_name=fake_return_arg_name)
+        assert not callback.future.done()
+        callback(return_arg=fake_return_arg_value)
+        await asyncio.sleep(0.1)  # wait to give time to complete the callback
+        assert callback.future.done()
+        assert not callback.future.exception()
+        assert await callback.completion() == fake_return_arg_value
 
-    @pytest.mark.it("Can be called using explicit keyword arguments")
-    async def test_can_be_called_using_explicit_kwargs(self, mocker, mock_function):
-        callback = async_adapter.AwaitableCallback(mock_function)
-        result = callback(a=1, b=2, c=3)
-        assert mock_function.call_count == 1
-        assert mock_function.call_args == mocker.call(a=1, b=2, c=3)
-        assert result == mock_function.return_value
+    @pytest.mark.it(
+        "Raises an exception when a call is invoked on the instance without the correct return argument (with return_arg_name)"
+    )
+    async def test_calling_object_raises_exception_if_return_arg_is_missing(
+        self, fake_return_arg_name, fake_return_arg_value
+    ):
+        callback = async_adapter.AwaitableCallback(return_arg_name=fake_return_arg_name)
+        with pytest.raises(TypeError):
+            callback()
 
-    @pytest.mark.it("Can have its callback completion awaited upon")
-    async def test_awaiting_completion_of_callback_returns_result(self, mock_function):
-        callback = async_adapter.AwaitableCallback(mock_function)
-        callback()
-        assert await callback.completion() == mock_function.return_value
+    @pytest.mark.it(
+        "Causes an error to be set on the instance Future when an error parameter is passed to the call (without return_arg_name)"
```

Comment:
I agree - writing "(without return_arg_name)" felt so wrong.  I've been moving away from parameterizing *args because I keep running into cases where different args lead to different tests and the copypasta was starting to hurt.

There's another way to do this, but I don't think this is any cleaner but it might be useful in some cases.  Instead of parametrizing *args, we could parametrize a lambda that called the function under test.

To illustrate, normally we do this:
```
@pytest.mark.parametrize(
    "arguments",
    [
        pytest.param((arg1), id="with 1 arg"),
        pytest.param((arg1, args), id="with 2 args"),
    ]
)
def test_blah(self, arguments):
    foo = ObjectToTest()
    foo.call_function(*arguments)
    assert foo.blah()
```

Instead, there _may_ be value in doing it this way:
```
@pytest.mark.parametrize(
    "invoke",
    [
        pytest.param(lambda obj: obj.call_function(arg1), id="with 1 arg"),
        pytest.param(lambda obj: obj.call_function(arg1, arg2), id="with 2 args"),
    ]
)
def test_blah(self, invoke):
    foo = ObjectToTest()
    invoke(foo)
    assert foo.blah()
```

This probably obfuscates, but I still think it's an interesting idea.

---
In reply to: [318840351](https://github.com/Azure/azure-iot-sdk-python-preview/pull/174#discussion_r318840351) [](ancestors = 318840351)